larpandoracontent/LArThreeDReco/LArCosmicRay/DeltaRayRemovalTool.cc

```diff
@@ -0,0 +1,232 @@
+       std::cout << "----> Running Algorithm Tool: " << this->GetInstanceName() << ", " << this->GetType() << std::endl;
+
+    bool changesMade(false);
+
+    ClusterVector sortedKeyClusters;
+    overlapTensor.GetSortedKeyClusters(sortedKeyClusters);
+
+    ClusterSet usedKeyClusters;
+    for (const Cluster *const pKeyCluster : sortedKeyClusters)
+    {
+        if (usedKeyClusters.count(pKeyCluster))
+            continue;
+
+        TensorType::ElementList elementList;
+        overlapTensor.GetConnectedElements(pKeyCluster, true, elementList);
+        
+        for (const TensorType::Element &element : elementList)
+            usedKeyClusters.insert(element.GetClusterU());
+
+        this->ExamineConnectedElements(elementList, changesMade);
+    }
+    
+    return changesMade;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void DeltaRayRemovalTool::ExamineConnectedElements(const TensorType::ElementList &elementList,
+    bool &changesMade) const
+{
+    ClusterSet modifiedClusters, checkedClusters;
+    
+    for (const TensorType::Element &element : elementList)
+    {
+        for (const HitType hitType : {TPC_VIEW_U, TPC_VIEW_V, TPC_VIEW_W})
+	    {
+            const Cluster *pDeltaRayCluster(element.GetCluster(hitType));
+            const ParticleFlowObject *const pMuonPfo(element.GetOverlapResult().GetCommonMuonPfoList().front());
+
+            if (checkedClusters.count(pDeltaRayCluster))
+                continue;
+            
+            if ((modifiedClusters.count(element.GetClusterU())) || (modifiedClusters.count(element.GetClusterV())) || (modifiedClusters.count(element.GetClusterW())))
+                continue;
+
+            if (!this->PassElementChecks(element, hitType))
+                continue;
+
+            if (!this->IsContaminated(element, hitType))
+                continue;            
+
+            if (!this->IsBestElement(element, hitType, elementList))
+                continue;
+            
+            checkedClusters.insert(pDeltaRayCluster);
+
+            CaloHitList deltaRayHits;
+            if (m_pParentAlgorithm->CollectHitsFromMuon(nullptr, nullptr, pDeltaRayCluster, pMuonPfo, m_minDistanceFromMuon, m_maxDistanceToCollected, deltaRayHits) != STATUS_CODE_SUCCESS)
+                continue;
+            
+            modifiedClusters.insert(pDeltaRayCluster);
+
+            this->SplitMuonCluster(element, hitType, deltaRayHits);
+         }
+    }
+
+    changesMade = modifiedClusters.size();
+}
+    
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool DeltaRayRemovalTool::PassElementChecks(const TensorType::Element &element, const HitType hitType) const
+{
+    // ATTN: Avoid endpoints, topological michel reconstruction is very ambiguous
+    if (this->IsMuonEndpoint(element, false))
```

Comment:
This appears to be the only difference relative to the CosmicRayRemovalTool version of this function. This function may be a candidate to move up to the RemovalBaseTool, either providing the boolean conditional as a parameter and working out which version of IsMuonEndpoint to call, or by moving the code below this point to a common (protected) RemovalBaseTool function that could be called from this function in each of the inheriting classes.

Reply:
i think i'll do the latter since the boolean seems a bit far removed to be a parameter to PassElementChecks